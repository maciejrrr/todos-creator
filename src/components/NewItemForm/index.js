import React, { Component } from 'react';
import { uniqueId } from 'lodash';
import styled, { css } from 'styled-components';

import FormActions from '../FormActions';
import { metrics, colors } from '../../theme';

const Form = styled.form`
  padding: ${props => (props.inline ? 0 : metrics.smallMargin)}px;
  width: ${metrics.formWidth}px;
  border-radius: ${metrics.borderRadius}px;
  background-color: ${props => (props.focused ? colors.cardBackground : 'transparent')};
  ${props =>
    props.inline &&
    css`
      display: flex;
      width: auto;
      flex-direction: row;
      align-items: center;
    `};
`;

const Input = styled.input`
  background-color: ${colors.addCardBackground};
  width: 100%;
  border: 0;
  outline-width: 0;
  height: ${metrics.inputHeight}px;
  padding: ${metrics.smallMargin}px;
  color: ${colors.addCardText};
  ::placeholder {
    color: ${colors.addCardText};
  }
  ${props =>
    props.focused &&
    css`
      color: ${colors.grayText};
      ::placeholder {
        color: ${colors.grayText};
      }
    `};
`;

class NewItemForm extends Component {
  state = {
    name: '',
    focused: false,
  };

  handleSubmit = e => {
    e.preventDefault();
    this.props.submit({
      item: {
        id: uniqueId(),
        name: this.state.name,
      },
    });
    this.handleResetForm();
    this.input.blur();
  };

  handleChange = e => {
    this.setState({ name: e.target.value });
  };

  handleFocus = () => {
    this.setState({ focused: true });
  };

  handleResetForm = () => {
    this.setState({ focused: false, name: '' });
  };

  render() {
    const { focused, name } = this.state;
    const { inline, placeholder } = this.props;
    return (
      <Form focused={focused} inline={inline} onSubmit={this.handleSubmit}>
        <Input
          innerRef={input => (this.input = input)}
          focused={focused}
          placeholder={placeholder}
          onChange={this.handleChange}
          value={name}
          onFocus={this.handleFocus}
        />
        <FormActions
          inline={inline}
          focused={focused}
          disabled={!name}
          text="Add"
          resetForm={this.handleResetForm}
        />
      </Form>
    );
  }
}

export default NewItemForm;
