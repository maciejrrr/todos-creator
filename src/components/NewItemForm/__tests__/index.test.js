import NewItemForm from '../index';

describe('NewItemForm component', () => {
  let props;
  beforeEach(() => {
    props = {
      submit: jest.fn(),
    };
  });

  describe('snapshots', () => {
    describe('initial state', () => {
      it('renders without focused components', () => {
        const tree = renderer.create(<NewItemForm {...props} />).toJSON();
        expect(tree).toMatchSnapshot();
      });
    });

    describe('focused state', () => {
      it('renders focused form', () => {
        const component = renderer.create(<NewItemForm {...props} />);
        component.getInstance().handleFocus();
        expect(component.toJSON()).toMatchSnapshot();
      });
    });

    describe('inline form', () => {
      it('renders inline form', () => {
        const component = renderer.create(<NewItemForm {...props} inline />);
        component.getInstance().handleFocus();
        expect(component.toJSON()).toMatchSnapshot();
      });
    });
  });

  describe('handleSubmit', () => {
    let e;
    let preventDefaultMock;
    let handleResetFormMock;
    let blurMock;
    let name;
    beforeEach(() => {
      preventDefaultMock = jest.fn();
      handleResetFormMock = jest.fn();
      blurMock = jest.fn();
      const component = shallow(<NewItemForm {...props} />);
      e = { preventDefault: preventDefaultMock };
      name = 'valid name';
      component.instance().handleResetForm = handleResetFormMock;
      component.instance().input = { blur: blurMock };
      component.setState({ name, focused: true });
      component.instance().handleSubmit(e);
    });

    it('calls preventDefault', () => {
      expect(preventDefaultMock).toHaveBeenCalledTimes(1);
    });

    it('calls submit', () => {
      expect(props.submit).toBeCalledWith({
        item: {
          id: 1,
          name,
        },
      });
    });

    it('calls handleResetForm', () => {
      expect(handleResetFormMock).toHaveBeenCalledTimes(1);
    });

    it('calls blur on input ref', () => {
      expect(blurMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('handleChange', () => {
    it('changes name state', () => {
      const component = shallow(<NewItemForm {...props} />);
      const value = 'test value';
      const e = { target: { value } };
      component.instance().handleChange(e);
    });
  });

  describe('handleFocus', () => {
    it('changes focused state', () => {
      const component = shallow(<NewItemForm {...props} />);
      component.instance().handleFocus();
      expect(component.state().focused).toBe(true);
    });
  });

  describe('handleResetForm', () => {
    let component;
    beforeEach(() => {
      component = shallow(<NewItemForm {...props} />);
      component.setState({ focused: true, name: 'test name' });
      component.instance().handleResetForm();
    });

    it('changes focused state', () => {
      expect(component.state().focused).toBe(false);
    });

    it('changes name state', () => {
      expect(component.state().name).toBe('');
    });
  });
});
