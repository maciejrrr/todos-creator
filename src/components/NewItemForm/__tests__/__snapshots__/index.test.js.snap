// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewItemForm component snapshots focused state renders focused form 1`] = `
<form
  className="sc-bxivhb bdIKGc"
  onSubmit={[Function]}
>
  <input
    className="sc-ifAKCX cRsMfw"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    value=""
  />
  <div
    className="sc-bdVaJa jIKszl"
  >
    <button
      className="sc-bwzfXH hGZcQI"
      disabled={true}
    >
      Add
    </button>
    <i
      className="anticon anticon-close sc-htpNat oPmsX"
      onClick={[Function]}
    />
  </div>
</form>
`;

exports[`NewItemForm component snapshots initial state renders without focused components 1`] = `
<form
  className="sc-bxivhb gAAXBk"
  onSubmit={[Function]}
>
  <input
    className="sc-ifAKCX cMqBtW"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    value=""
  />
</form>
`;

exports[`NewItemForm component snapshots inline form renders inline form 1`] = `
<form
  className="sc-bxivhb fSZRgy"
  onSubmit={[Function]}
>
  <input
    className="sc-ifAKCX cRsMfw"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    value=""
  />
  <div
    className="sc-bdVaJa leeTYR"
  >
    <button
      className="sc-bwzfXH hGZcQI"
      disabled={true}
    >
      Add
    </button>
    <i
      className="anticon anticon-close sc-htpNat oPmsX"
      onClick={[Function]}
    />
  </div>
</form>
`;
